### ISSUE TITLE
What serialization(s) will IEEE 3195.1 CCO be in?

### ISSUE THREAD

**Issue by jimschoening1** [CommonCoreOntology/CommonCoreOntologies#202]
Brian Haugh asked: "There is also a question of whether to use Turtle, RDF/XML, or other serializations of the CCO for the reference version. I believe that the DIOWG guidance is leaning towards requiring RDF/XML for ontologies in its Foundry."
Background:  When approved, IEEE 3195.1 CCO will be a printable document, but could also refer to a digital file (most likely stored as a given CCO version at https://opensource.ieee.org/cco).  The paper document will be available for sale as normal, and the digital file will be posted for free download. 
My Questions: What serializations will CCO be offered in?  If more than one, are they all equivalent, or is one the standard and others derived?  Will the printable document refer to the digital file?  If so, is that acceptable for a standard?  How did ISO/IEC-JTC1-21838-2 BFO handle this?

**Comment by cameronmore** [comment]
We discussed recently in a IEEE CCO subgroup meeting that we might want to reflect the BFO-ISO standard approach: the standard isn't the RDF file _per se_ but a written document that outlines _how_ portions of BFO satisfy the top-level requirements, and points to the GitHub as the site where it's developed and housed, so we should do the same thing with CCO. The particular version of CCO we submit will be changed, updated, expanded, and modified for as long as it's used and contributed to, so we shouldn't think of there being an "IEEE Official Version"

**The standard can of course include the file**, but the heart of the standard is the written explanation showing how the CCO satisfies mid-level ontology criteria. Moreover, the CCO is really a collection of modules, so the merged-all-core file is technically a convenience file more than a single grand ontology.

Theoretically, RDF, OWL, TTL should be equivalent, so it might not technically matter. However, as developers, we prefer TTL serialization far more than others because it's easier to read, easier to 'cut and paste,' easier to transfer into SPARQL and SHACL, and easier to teach people how to use/read. I personally advocate TTL all the way down, in DIOWG as well (I'm not sure how XML got suggested as the DIOWG standard).

I can gladly leave this issue open how a little while for others to read and offer thoughts. @mark-jensen @johnbeve

**Comment by johnbeve** [comment]
> My Questions: What serializations will CCO be offered in? If more than one, are they all equivalent, or is one the standard and others derived? 

RDF serializations are equivalent. Any one can be converted into any of the others, using publicly-available [tools](https://issemantic.net/rdf-converter) or library features [rdflib](https://rdflib.readthedocs.io/en/stable/plugin_serializers.html) (not endorsing either, just indicating examples). 

Reasons to prefer, say, RDF/XML over N-Triples or Turtle or JSON-LD will depend on implementation details, so neutrality is important. I'm partial to Turtle because I find it easy to read, but I've met many developers who prefer JSON-LD for the same reason. 

These observations suggest though that we should provide users with or direct users to directions for re-serialization strategies. 

> Will the printable document refer to the digital file? If so, is that acceptable for a standard? How did ISO/IEC-JTC1-21838-2 BFO handle this?

ISO 21838 can be [purchased in both printed and pdf+epub formats](https://www.iso.org/standard/71954.html), but it is also downloadable for free assuming you [agree to use the download to develop other ISO standards](https://www.iso.org/terms-conditions-licence-agreement.html#Customer-Licence). As I understand, the content in each is the same.

**Comment by jimschoening1** [comment]
I feel this issue is important enough to be fully vetted, presented to, and voted on by the Ontology Standards Working Group, so please leave this open until after a vote passes. 

CCO has the attached approved Project Authorization Request (PAR):
[P3195.1 (1).pdf](https://github.com/CommonCoreOntology/CommonCoreOntologies/files/13405850/P3195.1.1.pdf)

5.3 states: "This standard must conform to IEEE P3195 Standard for Requirements for a Mid-Level Ontology."   So, I agree CCO must show conformance.

5.2 states: "Scope of proposed standard: This standard defines a mid-level ontology that specifies a set of well-defined terms and relations commonly used across multiple domains..."   This states we are standardizing a specific ontology.  Let's say CCOv2.0 is what gets voted on and approved as a standard.  v2.1 can follow and be released online, but it won't be an IEEE standard unless later approved as such.   

I'm now starting to remember the unique arrangement Barry worked out with ISO, where they somehow delegated further development to his group of experts.  We can explore that, but I doubt it applies to CCO and is not what our PAR authorizes.  We would need to change the PAR to take that route. 

@johnbeve I suggest you try drafting the paragraph that references the approved standard CCO, so we can show it to people with experience in how to do this.   This might be simple, but I feel we should still get it reviewed sooner than later.

**Comment by jimschoening1** [comment]
The emerging Verifiable Credential ecosystem is using JSON-LD.  I see this ecosystem as needing P3195 ontologies and starting to use them as soon as the IEEE PURL server provides stable and trustworthy URLs.  If we standardize on one serialization, it doesn't need to be JSON-LD (provided we provide a version in JSON-LD)(so why don't we provide various versions now for CCO v1.4?).  As for PURL responses, why not give it in multiple formats, plus allow the query to ask for a given version.  This standard might help with this:  'Best Practice Recipes for Publishing RDF Vocabularies' https://www.w3.org/TR/swbp-vocab-pub/

**Comment by jimschoening1** [comment]
Let's consider this solution:  The 'normative' (i.e. what is a standard) wording in P3195.1 CCO (the document that can be printed) would state: "CCO ontology is standardized in the following equivalent serializations (Turtle Syntax, RDF/XML Syntax, OWL/XML Syntax, and JSON-LD) and available for download at <PURL>."  We could also include 'informative' wording to explain how additional equivalent serializations can be derived.

**Comment by johnbeve** [comment]
> CCO ontology is standardized in the following equivalent serializations (Turtle Syntax, RDF/XML Syntax, OWL/XML Syntax, and JSON-LD) and available for download at...

I don't think you should attempt to enumerate serialization formats, partly for the reason raised by @swartik. I suggest: "The CCO artifact is serialized in a standard RDF format at..."

We can then write brief guidance located in the README wherever "..." points, explaining that serializations are semantically equivalent and how users can re-serialize to another format if needed for their environment.

**Comment by jimschoening1** [comment]
John,     When drafting the PAR, we were severely criticized by an IEEE senior reviewer for not knowing the difference between an artifact and its serialization.  I hear consensus the core of this standard should be CCO.ttl.  I propose we don't call it a serialization, because that begs the question, 'Serialization of what artifact, and shouldn't that artifact be the standard?'  But there is no single artifact that could be the standard.     So, could the standard state: "This document standardizes MergedAllCoreOntology-v1.4-2023-04-07.ttl as found at https://opensource.ieee.org/cco/CommonCoreOntologies/-/blob/master/cco-merged/MergedAllCoreOntology-v1.4-2023-04-07.ttl." (Replace with correct version and PURL and list all ontology modules if we decide to. ) 

Also, I'm working with the emerging Verifiable Credential community, which uses JSON-LD.  Is there any reason why we could not offer CCO in multiple forms?

**Comment by cameronmore** [comment]
We could, but if you have Python and the RDFLib package, it only takes a moment, this code converts it:

```
from rdflib import Graph

g = Graph()

# Load the ontology with the file name
g.parse("CCO.ttl")

# Change the serialization format
g.serialize("CCO.jsonld", format="json-ld")

```

**Comment by jimschoening1** [comment]
But that could still be an obstacle for adoption.  Some might not believe it is equivalent or part of the standard.  So if there is no reason it couldn't be provided as a version of the standard, I propose it should.   Also, for JSON-LD, I committed to do a demo to the Open Wallet Foundation of how a Verifiable Credential claim (for a term, e.g. height) includes a PURL in its JSON-LD code that returns a definition.  As such, I'll need a version of CCO in JSON-LD, which I propose we load on our Gitlab repo.  Plus our PURL Server will need to respond to queries in Turtle, JSON-LD, and any other from we include.     Thoughts anyone?

**Comment by mark-jensen** [comment]
>  propose we don't call it a serialization, because that begs the question, 'Serialization of what artifact, and shouldn't that artifact be the standard?' 

I agree. This concern about whether the specific files named in the standard are serialized in turtle or xml, or some other,  is a false dilemma. We are standardizing the content of the 11 ontologies that comprise a particular named and dated release of CCO. It has to be verifiable OWL2 in _some_ standard serialization. Is there a reason we need to provide more than format for the standard? Where did that come from? In the ISO spec for BFO under section 4.3 "OWL 2 formalization of BFO-2020", it names two OWL versions, one in RDF (presumably this is XML) and the other in OWL functional syntax. I am not sure why they did this, maybe due to how BFO was built or some other requirement related to the FOL version or consistency checking. @alanruttenberg @johnbeve Do you recall why?

> Is there any reason why we could not offer CCO in multiple forms?

If there is enough of a demand signal from users to provide alternate encodings, then of course we can. 

> Plus our PURL Server will need to respond to queries in Turtle, JSON-LD, and any other from we include. 

Who decided that the PURL server has to negotiate gets for JSONLD? Or turtle for that matter? More importantly, what does that have to do with the format(s) of what gets submitted as the standard? 

It is important to keep these three things separate:
1. CCO as it exists in the repository. It will change as edits are made and new releases generated. 
2. The files that are returned via the IRIs. These are generated using those in no. 1, but nonetheless separate, and will always reflect the current release of CCO. 
3. The files that get named in the P3195.1. These will (I assume) be stored somewhere else as an archive where they can be accessed independently of the repo. They will be copies of the ontology files at certain point in the commit history of the repo, a snapshot of a version of CCO. 

@jimschoening1 How does IEEE SA store and archive standards once accepted?

I'd like to note that the scope of this issue has crept and I think it would be wise to keep it limited to the specifics of what   serialization(s) are named, if at all, in the standard 3195.1, as well as the language surrounding how they are described. Discussion about which formats are needed to be returned via the PURLs is a separate issue. Requests for providing CCO in alternate encodings for use-cases is another issue.

**Comment by swartik** [comment]
@cameronmore, are you suggesting your translation script be part of the CCO standard? I am wary of that. You would have to specify the versions of Python and rdflib. You would have to demonstrate that the script behaves equivalently across platforms and Python implementations. If I were an IEEE reviewer, I would look askance at anything less.

If you're just talking about what expert users can do, that's different. It still leaves open the question of what the standard should contain, and why.

@jimschoening1, can you educate us (me?) about serializations in standards? Is it common to include multiple serializations? What are popular serializations? What documentation describes the relationship between an artifact and its serialization? And what's the relationship between the standard and the gitlab server? Does the standard mention the gitlab server? And the PURL server, for that matter? Or the other way around? Or both? Do standards describe how to obtain serializations?

**Comment by cameronmore** [comment]
@swartik definitely not suggesting my script be part of the standard, just making it easier for someone to convert if they want.

**Comment by johnbeve** [comment]
> John, When drafting the PAR, we were severely criticized by an IEEE senior reviewer for not knowing the difference between an artifact and its serialization. 

Understood. I think we can display awareness of the difference without asserting a specific serialization. 

> I propose we don't call it a serialization, because that begs the question, 'Serialization of what artifact, and shouldn't that artifact be the standard?' But there is no single artifact that could be the standard.

Agreed. To be clear, my suggestion "The CCO artifact is serialized..." was meant to be compatible with the fact that: 
1. What is being standardized is (as @mark-jensen notes) the _content_ of the CCO ontologies 

While conveying that:

2. The content of CCO is represented in _some_ serialization format.
3. The standard doesn't dictate _what_ format.  

Today the serialization might be TTL; in three years it might be RDF/XML. 

> I'd like to note that the scope of this issue has crept and I think it would be wise to keep it limited to the specifics of what serialization(s) are named, if at all, in the standard 3195.1, as well as the language surrounding how they are described.

Agreed.

**Comment by jimschoening1** [comment]
I checked with Jonathn Goldberg, our IEEE Staff Advisor, who clarified our standards document will include a link to the ontology.  He said, " Yes, the OS repository is linked in the draft pointing to the code. The code is not included in the draft. "   I conclude this also means we can easily point to multiple serializations and assert they are equivalent.

**Comment by alanruttenberg** [comment]
If you look in the [functional syntax spec](https://www.w3.org/TR/owl-ref/#MIMEType) you will see 

>This document also defines the functional-style syntax, which closely follows the structural specification and allows OWL 2 ontologies to be written in a compact form. This syntax is used in the definitions of the semantics of OWL 2 ontologies, the mappings from and into the **RDF/XML exchange** syntax, and the different profiles of OWL 2. Concrete syntaxes, such as the functional-style syntax, often provide features not found in the structural specification, such as a mechanism for abbreviating IRIs"

All the files mentioned in that document have the extension .owl. It also says:
> It is recommended that OWL functional-style Syntax files have the extension .ofn (all lowercase) on all platforms. 

The [OWL/XML document](https://www.w3.org/TR/owl2-xml-serialization/) says
>It is recommended that OWL XML Serialization files have the extension .owx (all lowercase) on all platforms. 

The [Manchester syntax document](https://www.w3.org/TR/owl2-manchester-syntax/) says:
>It is recommended that OWL Manchester Syntax files have the extension ".omn" (all lowercase) on all platforms. 

If you save an ontology from Protege without giving a file extension OWL/XML will get the extension .owx. If you save it as functional syntax it will get the extension .ofn . If you save it in the Manchester syntax you will get .omn.

Practically speaking you can save all of these with ".owl" and RDF/XML with ".rdf". The [OWL 1 reference](https://www.w3.org/TR/owl-ref/#MIMEType) says uses .owl or .rdf. However, when I see .rdf I always am not sure whether it is OWL or generic RDF, so I don't use that and dislike when others do for owl, given that there are established conventions.

The RDF/XML document does say
>It is recommended that RDF/XML files have the extension ".rdf" (all lowercase) on all platforms. 

Protege will save RDF/XML with a default .rdf extension, However this is not the convention generally used for OWL files, c.f. the functional syntax document quote above.

The parsers I and Protege use try to figure out the file format based on the contents without really consulting the file extension.

I don't like content negotiation because it takes server configuration to work and because you can't be sure what you are going to get if you put the URL into a browser. I consider the different file formats to be distinct resources, though opinions differ.

**Comment by swartik** [comment]
@alanruttenberg  Thanks. I haven't looked at the OWL 1 specifications for a long time. Nor have I memorized all the OWL 2 documentation. It always helps when one of the authors contributes.

That said, I'm not convinced suffix used is far enough along to qualify as having widely accepted established conventions. I had a sponsor tell me his tool required an RDF/XML file to have a name end with ".xml". I hope he's an outlier.

What version of Protege are you using? I'm running 5.5.0 on Windows. I just tried saving an ontology in RDF/XML, OWL/XML, and Turtle by entering a non-suffixed name into the file name box. In all three cases, Protege added ".owl" as the suffix.
